core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java

```diff
@@ -184,25 +195,73 @@ private void copyBytes(Block block, int[] lengths, int[] positions, int count, i
     }
 
     /**
-     * Copy {@code length} bytes from {@code block}, at position {@code position} to {@code count} consecutive positions in the {@link #bytes} array.
+     * Copy {@code length} bytes from {@code slice}, starting at offset {@code sourceOffset} to {@code count} consecutive positions in the {@link #bytes} array.
      */
-    private void duplicateBytes(Block block, int position, int count)
+    private void duplicateBytes(Slice slice, int sourceOffset, int length, int count)
     {
-        int length = block.getSliceLength(position);
         int newByteCount = toIntExact((long) count * length);
         int startOffset = getCurrentOffset();
-        ensureBytesCapacity(startOffset + newByteCount);
+        ensureExtraBytesCapacity(newByteCount);
 
-        Slice slice = block.getSlice(position, 0, length);
+        if (length > 0) {
+            duplicateBytes(slice, sourceOffset, bytes, startOffset, length, count);
```

Comment:
`if (length > 0) {` check is redundant (there is check already in `duplicateBytes`)